test/extensions/filters/http/ext_authz/ext_authz_test.cc

```diff
@@ -152,38 +184,133 @@ class HttpFilterTest : public HttpFilterTestBase<testing::Test> {
+    response_.status = status_ok_ ? Filters::Common::ExtAuthz::CheckStatus::OK
+                                  : Filters::Common::ExtAuthz::CheckStatus::Denied;
+  }
 
-  const std::string grpc_config = R"EOF(
-  transport_api_version: V3
-  grpc_service:
-    transport_api_version: V3
-    envoy_grpc:
-      cluster_name: "ext_authz_server"
-  )EOF";
+  static std::string ParamsToString(const testing::TestParamInfo<bool>& info) {
+    return info.param ? "StatusOk" : "StatusNotOk";
+  }
 
-  envoy::extensions::filters::http::ext_authz::v3::ExtAuthz proto_config{};
-  TestUtility::loadFromYaml(http_client ? http_config : grpc_config, proto_config);
-  proto_config.set_failure_mode_allow(failure_mode_allow_value);
-  return proto_config;
-}
+  void testResponse(TestResponseOpts opts = {}) {
+    InSequence s;
 
-INSTANTIATE_TEST_SUITE_P(ParameterizedFilterConfig, HttpFilterTestParam,
-                         Values(&getFilterConfig<true, true>, &getFilterConfig<false, false>,
-                                &getFilterConfig<true, false>, &getFilterConfig<false, true>));
+    initialize(R"(
+        transport_api_version: V3
+        grpc_service:
+          envoy_grpc:
+            cluster_name: "ext_authz_server"
+        failure_mode_allow: false
+    )");
+
+    // Simulate a downstream request.
+    ON_CALL(decoder_filter_callbacks_, connection())
+        .WillByDefault(Return(OptRef<const Network::Connection>{connection_}));
+    connection_.stream_info_.downstream_connection_info_provider_->setRemoteAddress(addr_);
+    connection_.stream_info_.downstream_connection_info_provider_->setLocalAddress(addr_);
+    EXPECT_CALL(*client_, check(_, _, _, _))
+        .WillOnce(Invoke(
+            [&](Filters::Common::ExtAuthz::RequestCallbacks& callbacks,
+                const envoy::service::auth::v3::CheckRequest&, Tracing::Span&,
+                const StreamInfo::StreamInfo&) -> void { request_callbacks_ = &callbacks; }));
+
+    EXPECT_EQ(Http::FilterHeadersStatus::StopAllIterationAndWatermark,
+              filter_->decodeHeaders(request_headers_, false));
+
+    if (status_ok_) {
+      testUpstreamRequest(opts.remove_downstream_req_header, opts.add_authz_req_header);
+      testEncodeHeaders(opts.add_authz_resp_header);
+    } else {
+      testLocalReply(opts.add_authz_local_reply_header);
+    }
+  }
+
+  void testUpstreamRequest(bool remove_downstream_req_header, bool add_authz_req_header) {
+    // Add a header to see if it gets removed later.
+    request_headers_.addCopy(downstream_req_header_name_, downstream_req_header_value_);
+    EXPECT_EQ(Http::FilterHeadersStatus::StopAllIterationAndWatermark,
+              filter_->decodeHeaders(request_headers_, false));
+    EXPECT_CALL(decoder_filter_callbacks_, continueDecoding());
+
+    request_callbacks_->onComplete(
+        std::make_unique<Filters::Common::ExtAuthz::Response>(response_));
+
+    // Now make sure the downstream header is / is not there depending on the test.
```

Comment:
Do you need to add a test to verify the invalid headers {"invalid-\nkey", "bar"} is not added? Or you already have such test somewhere?